l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,742 @@
+            );
+    }
+
+    struct MessageParams {
+        uint256 l2BatchNumber;
+        uint256 l2MessageIndex;
+        uint16 l2TxNumberInBatch;
+    }
+
+    function _finalizeWithdrawal(
+        uint256 _chainId,
+        uint256 _l2BatchNumber,
+        uint256 _l2MessageIndex,
+        uint16 _l2TxNumberInBatch,
+        bytes calldata _message,
+        bytes32[] calldata _merkleProof
+    ) internal nonReentrant returns (address l1Receiver, address l1Token, uint256 amount) {
+        require(!isWithdrawalFinalizedShared[_chainId][_l2BatchNumber][_l2MessageIndex], "pw2");
+        isWithdrawalFinalizedShared[_chainId][_l2BatchNumber][_l2MessageIndex] = true;
+
+        if ((_chainId == ERA_CHAIN_ID) && ((_l2BatchNumber < eraIsEthWithdrawalFinalizedStorageSwitchBatchNumber))) {
+            // in this case we have to check we don't double withdraw ether
+            // we are not fully finalized if eth has not been withdrawn
+            // note the WETH bridge has not yet been deployed, so it cannot be the case that we withdrew Eth but not WETH.
+            bool alreadyFinalized = IGetters(ERA_DIAMOND_PROXY).isEthWithdrawalFinalized(
+                _l2BatchNumber,
+                _l2MessageIndex
+            );
+            require(!alreadyFinalized, "Withdrawal is already finalized");
+        }
+
+        bool wrapToWeth;
+
+        {
+            MessageParams memory messageParams = MessageParams({
+                l2BatchNumber: _l2BatchNumber,
+                l2MessageIndex: _l2MessageIndex,
+                l2TxNumberInBatch: _l2TxNumberInBatch
+            });
+
+            (l1Receiver, l1Token, amount, wrapToWeth) = _checkWithdrawal(
+                _chainId,
+                messageParams,
+                _message,
+                _merkleProof
+            );
+        }
+
+        if (!hyperbridgingEnabled[_chainId]) {
+            // Check that the chain has sufficient balance
+            require(chainBalance[_chainId][l1Token] >= amount, "EB n funds 2"); // not enought funds
+            chainBalance[_chainId][l1Token] -= amount;
+        }
+
+        if (wrapToWeth) {
+            // Wrap ETH to WETH tokens (smart contract address receives the equivalent amount of WETH)
+            IWETH9(l1WethAddress).deposit{value: amount}();
+            // Transfer WETH tokens from the smart contract address to the withdrawal receiver
+            IERC20(l1WethAddress).safeTransfer(l1Receiver, amount);
+
+            emit EthWithdrawalFinalized(_chainId, l1Receiver, amount);
+        } else if ((l1Token == ETH_TOKEN_ADDRESS) || (l1Token == l1WethAddress)) {
+            bool callSuccess;
+            // Low-level assembly call, to avoid any memory copying (save gas)
+            assembly {
+                callSuccess := call(gas(), l1Receiver, amount, 0, 0, 0, 0)
+            }
+            require(callSuccess, "L1WB: withdraw failed");
+            emit EthWithdrawalFinalized(_chainId, l1Receiver, amount);
```

Comment:
are you sure that the same event should be emitted regardless of whether the ETH was wrapped?

Reply:
I am a bit fuzzy on what events are used for exactly. If we want to track the balance of tokens, and we use `EthWithdrawalFinalized` to track eth balance, then it should be emitted, because ether is leaving the system.

Reply:
Yeah, I am not sure either

Reply:
I don't think Weth withdrawing happened before, so I would treat it as Eth withdrawals, we just send it to the receiver via the Weth contract. But from our point of view that does not matter